torch/csrc/jit/serialization.h

```diff
@@ -0,0 +1,285 @@
+//                 to correctly identify the footer
+// [8 bytes] Offset of last record - The last record in this format is used
+//                                   as an index into the rest of the file, so
+//                                   a reader can use this offset to seek to
+//                                   the last record and read the index.
+// [56 bytes] Pad/reserved - Pad out the footer s.t. the whole file's size is a
+//                           multiple of 64 bytes.
+//
+//
+// When developing this format we want to pay particular attention to the
+// following use cases:
+//
+// -- Reading --
+// 1) Reading with full random access
+//   a) Reading with file api's such as fread()
+//   b) mmaping the file and jumping around the mapped region
+// 2) Reading with 1-pass sequential access
+//      -> A reader will need to build up a data structure of parsed structures
+//         as it reads
+//
+// -- Writing --
+// 1) Writing with full random access
+// 2) Writing with 1-pass sequential access
+//      -> We must take care not to require updating values that have already
+//         been written. We place the variable-length index at the end and do
+//         not put any indicies into the header to fulfill this constraint.
+
+namespace {
+  struct RecordTags {
+    enum {
+      STORAGE = 1,
+      FOOTER = 2,
+    };
+  };
+
+  // Common constants
+  static constexpr uint64_t kFileMagicNumber = 0x314843524f545950L; // PYTORCH1
+  static constexpr uint64_t kFieldAlignment = 64L; // 64 byte alignment supports up to AVX512 for mmap
+
+  // Reader-specific constants
+  static constexpr uint64_t kMaxSupportedFileFormatVersion = 0x1L;
+
+  // Writer-specific constants
+  static constexpr uint64_t kFileFormatVersion = 0x1L;
+  static constexpr uint8_t kPadValue = 0xEF;
+
+  void wrapPErrorAndThrow(const std::string& msg) {
+    std::ostringstream oss;
+    oss << msg << " : " << strerror(errno);
+    throw std::runtime_error(oss.str());
+  }
+}  // namespace
+
+class PyTorchFileReader {
+ public:
+  PyTorchFileReader(std::string filename) {
+    fp = std::fopen(filename.c_str(), "rb");
+    if (!fp) {
+      wrapPErrorAndThrow("Couldn't open file for reading!");
+    }
+    // Store file size so we know when we're done reading because the f* APIs
+    // don't do a good job of that
+    std::fseek(fp, 0L, SEEK_END);
+    file_size = std::ftell(fp);
+    std::fseek(fp, 0L, SEEK_SET);
+    readAndValidateFileHeader();
+    // Do this now since we're reasonably sure this is actually a PyT file from
+    // the header.
+    if (file_size % kFieldAlignment != 0) {
+      throw std::runtime_error("File length is not a multiple of the alignment"
+                               " size. Is this a valid PyTorch file?");
+    }
+    readAndValidateFileFooter();
+  }
+  std::string getLastRecord() {
+    return getRecordWithKey(last_record_offset);
+  }
+  std::string getRecordWithKey(uint64_t key) {
```

Comment:
std::string will make it impossible to change this to an mmap-based implementation later. If it returned `std::shared_ptr<void>` then we  would be able to change it to one later.

Reply:
I changed this while fucking around with pybind, i'll change it back